### PR TITLE
Add sudo for admin and support users

### PR DESCRIPTION
**Pull request by zaucker**
- Users with role `admin` or `support` may sudo to another account
- Upon logout after a sudo the old login is restored

### CODE REVIEW

File: lib/Agrammon/Web/Routes.pm6 [hunk 1]
```diff
@@ -458,16 +458,14 @@ sub application-routes(Agrammon::Web::Service $ws) {
...
-                        user       => $username,
-                        role       => $user.role.name,
-                        last_login => $user.last-login,
-                        news       => Nil,
-                        :!sudoUser,
+                        :$username,
+                        :role($user.role.name),
+                        :last_login($user.last-login),
+                        :news(Nil),
+                        :sudoUser($user.sudo-user),
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:468]: Fixed (came from an inconsistent frontend naming).

File: lib/Agrammon/Web/SessionUser.pm6 [hunk 2]
```diff
@@ -26,4 +44,8 @@ class Agrammon::Web::SessionUser is Agrammon::DB::User does Cro::HTTP::Auth {
     method from-json((:$logged-in = False, :$username = Str)) {
         self.new(:$logged-in, :$username).load
     }
+
+    method may-sudo {
+        return self.role.name eq 'admin' or ~self.role.name eq 'support';
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/SessionUser.pm6:49]: > Also, `self.role.name eq 'admin' | 'support'` :-)

Ah, sure, one of the Raku niceties :-)

[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/SessionUser.pm6:49]: > Why the `~` here? `eq` will stringify anyway.

Not sure where that came from. Didn't make any sense to have it on one of the two occurences.

File: lib/Agrammon/Web/SessionUser.pm6 [hunk 3]
```diff
@@ -4,19 +4,37 @@ use Agrammon::DB::User;
...
+    has Str $.sudo-username;
 
-    method auth($username, $password) {
-        $!logged-in = self.password-is-valid($username, $password);
+    method auth($username, $password, $sudo-username?) {
+        if $sudo-username {
+            die X::Agrammon::DB::User::MayNotSudo.new(:username($sudo-username)) unless self.may-sudo;
 
-        die X::Agrammon::DB::User::InvalidPassword.new unless $!logged-in;
+            $!logged-in = True;
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/SessionUser.pm6:14]: Yes, we didn't even get into `auth()` if we weren't logged in. Removed.

File: lib/Agrammon/Web/SessionUser.pm6 (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/SessionUser.pm6:49]: > And perhaps check they really are logged in.

Added a check.

File: lib/Agrammon/Web/Routes.pm6 [hunk 4]
```diff
@@ -458,16 +458,14 @@ sub application-routes(Agrammon::Web::Service $ws) {
 
         # TODO: implement news in auth()
         post -> Agrammon::Web::SessionUser $user, 'auth' {
-            request-body -> %data {
-                my $username = %data<user>;
-                my $password = %data<password>;
-                if $user.auth($username, $password) {
+            request-body -> (:$username, :$password, :sudoUsername($sudo-username), *%rest) {
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:461]: `auth()` is used as route for "normal" login and for sudo. The frontend supplied username is **not** used in `SessionUser::auth()` in the sudo case, though.

[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:461]: Fixed, setting from the logged-in user's username.

File: lib/Agrammon/Web/SessionUser.pm6 [hunk 5]
```diff
@@ -4,19 +4,37 @@ use Agrammon::DB::User;
...
 
         self.set-username($username);
         self.load;
         return self;
     }
 
     method logout() {
-        $!logged-in = False;
+        if $!sudo-user {
+            self.set-username($!sudo-username);
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Web/SessionUser.pm6:30]: No changed needed here, as handled in setting the return name inside the route to the logged-in user.

**Comment by zaucker** [comment]
@jnthn The logout process isn't working yet. When the `logout` route is called, the previously stored `sudo-username` attribute of `SessionUser` is undefined. What am I missing?

**Review by jnthn — APPROVED** [review]

**Pull request merged**